### PR TITLE
Master website portal fwi

### PR DESCRIPTION
**Pull request by fwintjens**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_portal/controllers/main.py [hunk 1]
```diff
@@ -55,7 +55,8 @@ def _get_archive_groups(self, model, domain=None, fields=None, groupby="create_d
     @http.route(['/my', '/my/home'], type='http', auth="user", website=True)
     def account(self, **kw):
         values = self._prepare_portal_layout_values()
-        return request.render("website_portal.portal_my_home", values)
+        response = request.render("website_portal.portal_my_home", values)
```
Comment (tde-banana-odoo) [inline addons/website_portal/controllers/main.py]: Reads like an unnecessary change to me ;) . I suppose some leftover of code debugging.

File: addons/website_portal_purchase/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+{
+    'name': 'Website Portal for Purchases',
+    'category': 'Website',
+    'summary': 'Add your purchase orders in the frontend portal',
+    'version': '1.0',
+    'description': """
+Add your purchase orders in the frontend portal. Your customers will be able to connect to their portal to see the list (and the state) of their purchase orders.
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/__manifest__.py]: Display purchase orders on your website. Customers will be able to access their purchase orders in their own portal, see their state and communicate with salesmen.

File: addons/website_portal_purchase/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    @http.route(['/my/purchase/<int:purchase>'], type='http', auth="user", website=True)
+    def portal_my_purchase_order(self, purchase=None, **kw):
+        purchase = request.env['purchase.order'].browse([purchase])
+        try:
+            purchase.check_access_rights('read')
+            purchase.check_access_rule('read')
+        except AccessError:
+            return request.render("website.403")
+        return request.render("website_portal_purchase.portal_my_purchase_order", {
+            'purchase': purchase.sudo(),
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/controllers/main.py]: Why sudoing if the current user has read access ?

File: addons/website_portal_purchase/security/ir.model.access.csv [hunk 4]
```diff
@@ -0,0 +1,12 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_purchase_order,purchase.order,purchase.model_purchase_order,base.group_portal,1,0,0,0
+access_purchase_order_line,purchase.order.line,purchase.model_purchase_order_line,base.group_portal,1,0,0,0
+access_product_list,product.pricelist,product.model_product_pricelist,base.group_portal,1,0,0,0
+access_res_partner,res.partner,base.model_res_partner,base.group_portal,1,0,0,0
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/security/ir.model.access.csv]: Nopenopenope !

File: addons/website_portal_purchase/security/ir.model.access.csv [hunk 5]
```diff
@@ -0,0 +1,12 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_purchase_order,purchase.order,purchase.model_purchase_order,base.group_portal,1,0,0,0
+access_purchase_order_line,purchase.order.line,purchase.model_purchase_order_line,base.group_portal,1,0,0,0
+access_product_list,product.pricelist,product.model_product_pricelist,base.group_portal,1,0,0,0
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/security/ir.model.access.csv]: id incoherent with model

Is it necessary to give access to pricelists to portal users ? Seems louche.

File: addons/website_portal_purchase/security/ir.model.access.csv [hunk 6]
```diff
@@ -0,0 +1,12 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_purchase_order,purchase.order,purchase.model_purchase_order,base.group_portal,1,0,0,0
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/security/ir.model.access.csv]: No restriction ?

File: addons/website_portal_purchase/security/ir.model.access.csv [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+access_purchase_order_line,purchase.order.line,purchase.model_purchase_order_line,base.group_portal,1,0,0,0
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/security/ir.model.access.csv]: No restriction ?

File: addons/website_portal_purchase/controllers/main.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        # search the purchase orders to display, according to the pager data
+        purchases = PurchaseOrder.search(
+            domain,
+            order=order,
+            limit=self._items_per_page,
+            offset=pager['offset']
+        )
+        values.update({
+            'date': date_begin,
+            'purchases': purchases,
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/controllers/main.py]: I would use orders. Basic object is a purchase order.

File: addons/website_portal_sale/controllers/main.py [hunk 9]
```diff
@@ -83,7 +81,7 @@ def portal_my_quotes(self, page=1, date_begin=None, date_end=None, sortby=None,
 
         values.update({
             'date': date_begin,
-            'quotations': quotations,
+            'quotations': quotations.sudo(),
```
Comment (tde-banana-odoo) [inline addons/website_portal_sale/controllers/main.py]: I suppose we spoke about that, but it seems strange.

File: addons/website_portal_sale/controllers/main.py [hunk 10]
```diff
@@ -132,7 +130,7 @@ def portal_my_orders(self, page=1, date_begin=None, date_end=None, sortby=None,
 
         values.update({
             'date': date_begin,
-            'orders': orders,
+            'orders': orders.sudo(),
```
Comment (tde-banana-odoo) [inline addons/website_portal_sale/controllers/main.py]: I suppose we spoke about that, but it seems strange.

File: addons/website_quote/controllers/main.py [hunk 11]
```diff
@@ -31,6 +31,10 @@ def view(self, order_id, pdf=None, token=None, message=False, **post):
         if not Order:
             return request.render('website.404')
 
+        # Token or not, sudo the order, since portal user has not access on
+        # taxes, required to compute the total_amout of SO.
+        Order = Order.sudo()
```
Comment (tde-banana-odoo) [inline addons/website_quote/controllers/main.py]: If sudoing is necessary, variable name should be changed in order to avoid confusion.

Comment (jem-odoo) [inline addons/website_quote/controllers/main.py]: ok

File: addons/website_portal_sale/controllers/main.py (hunk 10 above)
Comment (jem-odoo) [inline addons/website_portal_sale/controllers/main.py]: To display the total, it required to have read access to product.pricelist, account.tax, ...
AL does not want portal to have access to pricelist.
Also, on the frontend form view, portal needs to list the product.product name of its SO lines, and craque dedans ! Donc sudo is requis ;)

File: addons/website_portal_purchase/controllers/main.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+from odoo import http
+from odoo.exceptions import AccessError
+from odoo.http import request
+from odoo.tools.translate import _
+
+from odoo.addons.website_portal.controllers.main import website_account, get_records_pager
+
+
+class WebsiteAccount(website_account):
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/controllers/main.py]: I know we are crazy. But could'nt we put this definition in a website_portal.py file, and name the class WebsitePortal ?

Yo crazy moma !

File: addons/website_portal_purchase/controllers/main.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+    def _prepare_portal_layout_values(self):
+        values = super(WebsiteAccount, self)._prepare_portal_layout_values()
+        partner = request.env.user.partner_id
+        values['purchase_count'] = request.env['purchase.order'].search_count([
+            '|',
+            ('message_partner_ids', 'child_of', [partner.commercial_partner_id.id]),
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/controllers/main.py]: Just asking, isn't it already taken into account by access rules ?

File: addons/website_portal_purchase/controllers/main.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+        values.update({
+            'date': date_begin,
+            'orders': orders,
+            'page_name': 'purchase',
+            'pager': pager,
+            'archive_groups': archive_groups,
+            'searchbar_sortings': searchbar_sortings,
+            'sortby': sortby,
+            'searchbar_filters': OrderedDict(sorted(searchbar_filters.items())),
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/controllers/main.py]: Couldn't we directly create it accordingly ?

Is it intended to order it like 'all, cancel, done, purchase', or did I miss something ?

File: addons/website_portal_purchase/models/purchase.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class PurchaseOrder(models.Model):
+
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/models/purchase.py]: An evil kitten from hell arise.

File: addons/website_portal_purchase/models/purchase.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    _inherit = 'purchase.order'
+
+    website_url = fields.Char('Website URL', compute='_website_url', help='The full URL to access the document through the website.')
+
+    def _website_url(self):
+        for order in self:
+            order.website_url = '/my/purchase/%s' % (order.id)
+
+
+class PurchaseOrder(models.Model):
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/models/purchase.py]: Class name is not spec ready Sir able to.

File: addons/website_portal_purchase/models/purchase.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+    _inherit = 'purchase.order.line'
+
+    # Non-stored related field to allow portal user to see the image of the product he has ordered
+    product_image = fields.Binary('Product Image', related="product_id.image", store=False)
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/models/purchase.py]: Isn't store=False default value for related fields ?

File: addons/website_portal_purchase/security/website_portal_purchase_security.xml [hunk 18]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="1">
+
+        <record id="portal_purchase_order_user_rule" model="ir.rule">
+            <field name="name">Portal Purchase Orders</field>
+            <field name="model_id" ref="purchase.model_purchase_order"/>
+            <field name="domain_force">[('message_partner_ids','child_of',[user.commercial_partner_id.id])]</field>
+            <field name="groups" eval="[(4, ref('base.group_portal'))]"/>
+            <field eval="1" name="perm_unlink"/>
```
Comment (tde-banana-odoo) [inline addons/website_portal_purchase/security/website_portal_purchase_security.xml]: Non non non non non non. field name="my_name" eval="True/False" . Otherwise an evil kitten from hell with heavy laser beams will arise.

File: addons/website_portal_sale/controllers/main.py (hunk 10 above)
Comment (tde-banana-odoo) [inline addons/website_portal_sale/controllers/main.py]: If crak dedans then ok .

File: addons/website_portal_sale/controllers/main.py (hunk 9 above)
Comment (tde-banana-odoo) [inline addons/website_portal_sale/controllers/main.py]: Seems ok then, otherwise crak dedans.

File: addons/website_crm_partner_assign/security/ir_rule.xml [hunk 19]
```diff
@@ -1,10 +1,31 @@
...
+        <!-- Partner Grade : portal and public -->
+        <record id="res_partner_grade_rule_portal" model="ir.rule">
+            <field name="name">Portal user: read only website published</field>
+            <field name="model_id" ref="website_crm_partner_assign.model_res_partner_grade"/>
+            <field name="groups" eval="[(4, [ref('base.group_portal')])]"/>
+            <field name="domain_force">[('website_published','=', True)]</field>
+            <field name="perm_read" eval="True"/>
+        </record>
+
+        <record id="res_partner_grade_rule_public" model="ir.rule">
```
Comment (tde-banana-odoo) [inline addons/website_crm_partner_assign/security/ir_rule.xml]: Could be merged with res_partner_grade_rule_portal I think.

File: addons/website_crm_partner_assign/models/crm_lead.py [hunk 20]
```diff
@@ -190,17 +190,15 @@ def search_geo_partner(self):
...
     @api.multi
     def partner_interested(self, comment=False):
-        self.check_access_rights('write')
         message = _('<p>I am interested by this lead.</p>')
         if comment:
             message += '<p>%s</p>' % comment
         for lead in self:
             lead.message_post(body=message, subtype="mail.mt_note")
-            lead.sudo().convert_opportunity(lead.partner_id.id)
+            lead.sudo().convert_opportunity(lead.partner_id.id)  # sudo required to convert partner data
```
Comment (jem-odoo) [inline addons/website_crm_partner_assign/models/crm_lead.py]: Good catch ! Seems that it is required only for the `write` since the `write` needs the `read` to wins against the access rights. Or something like that !

**Comment by jem-odoo** [comment]
Landed in master. As it contains many none related feature, there are some commits : 

Community 
https://github.com/odoo/odoo/commit/936b215adffd2cb58c0c440d28dfe10adbffffe5
https://github.com/odoo/odoo/commit/25a067b58d12ed437b61829dbca0691e90d58bde
https://github.com/odoo/odoo/commit/ef75f0245f002cada1028cb6e1e7953858a019ba
https://github.com/odoo/odoo/commit/b90a8501b6263ff2e38b1cff29583872afc37439
